File: internal/features/project/service.go
package project

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"

	"jotter/backend/internal/features/common"
)

// Standard domain errors
var (
	ErrProjectNotFound = errors.New("project not found")
	ErrDuplicateProject = errors.New("project already exists")
	ErrLastProject      = errors.New("cannot delete the last remaining project")
	ErrInvalidInput     = errors.New("invalid input")
)

// Service coordinates domain business logic for projects
type Service interface {
	GetProjects(ctx context.Context) ([]Response, error)
	CreateProject(ctx context.Context, tasksDir string, req Create, defaultBuckets []map[string]interface{}, syncBucketsFunc func(string, string) error) (*Response, error)
	UpdateProject(ctx context.Context, tasksDir string, projectID string, raw map[string]interface{}, req Update) (*Response, error)
	DeleteProject(ctx context.Context, tasksDir string, projectID string) error
}

type projectService struct {
	dbRepo   DBRepository
	fileRepo FileRepository
}

// NewService creates a new project service instance
func NewService(dbRepo DBRepository, fileRepo FileRepository) Service {
	return &projectService{
		dbRepo:   dbRepo,
		fileRepo: fileRepo,
	}
}

func (s *projectService) GetProjects(ctx context.Context) ([]Response, error) {
	return s.dbRepo.GetAll(ctx)
}

func (s *projectService) CreateProject(ctx context.Context, tasksDir string, req Create, defaultBuckets []map[string]interface{}, syncBucketsFunc func(string, string) error) (*Response, error) {
	projectID := common.Slugify(req.Title)
	if projectID == "" {
		return nil, fmt.Errorf("%w: invalid title, could not generate a project ID slug", ErrInvalidInput)
	}

	projects, err := s.fileRepo.LoadProjects(tasksDir)
	if err != nil {
		return nil, fmt.Errorf("failed to load projects: %w", err)
	}

	for _, p := range projects {
		if p["id"] == projectID {
			return nil, fmt.Errorf("%w: project with ID '%s' already exists", ErrDuplicateProject, projectID)
		}
	}

	nowStr := time.Now().UTC().Format(time.RFC3339Nano)
	nowStr = strings.Replace(nowStr, "+00:00", "Z", 1)

	var gitRemoteVal interface{}
	if req.GitRemote != nil {
		gitRemoteVal = *req.GitRemote
	} else {
		gitRemoteVal = nil
	}

	newProject := map[string]interface{}{
		"id":                projectID,
		"title":             req.Title,
		"created_at":        nowStr,
		"done_clean_period": req.DoneCleanPeriod,
		"git_remote":        gitRemoteVal,
	}

	// Update Disk File
	projects = append(projects, newProject)
	if err := s.fileRepo.WriteProjects(tasksDir, projects); err != nil {
		return nil, fmt.Errorf("failed to write projects registry to disk: %w", err)
	}

	// Update SQLite database (including default buckets within internal database transaction)
	if err := s.dbRepo.Create(ctx, projectID, req.Title, nowStr, req.DoneCleanPeriod, req.GitRemote, defaultBuckets); err != nil {
		// Rollback file addition (by reloading, filtering, and writing back) to ensure transactional parity
		if rolledBackProjects, errReload := s.fileRepo.LoadProjects(tasksDir); errReload == nil {
			var filtered []map[string]interface{}
			for _, p := range rolledBackProjects {
				if p["id"] != projectID {
					filtered = append(filtered, p)
				}
			}
			_ = s.fileRepo.WriteProjects(tasksDir, filtered)
		}
		return nil, fmt.Errorf("failed to sync database: %w", err)
	}

	// Perform buckets file sync on disk
	if syncBucketsFunc != nil {
		_ = syncBucketsFunc(tasksDir, projectID)
	}

	return &Response{
		ID:              projectID,
		Title:           req.Title,
		CreatedAt:       nowStr,
		DoneCleanPeriod: req.DoneCleanPeriod,
		GitRemote:       req.GitRemote,
	}, nil
}

func (s *projectService) UpdateProject(ctx context.Context, tasksDir string, projectID string, raw map[string]interface{}, req Update) (*Response, error) {
	projects, err := s.fileRepo.LoadProjects(tasksDir)
	if err != nil {
		return nil, fmt.Errorf("failed to load projects: %w", err)
	}

	var projectFound map[string]interface{}
	for _, p := range projects {
		if p["id"] == projectID {
			projectFound = p
			break
		}
	}

	if projectFound == nil {
		return nil, fmt.Errorf("%w: project with ID '%s'", ErrProjectNotFound, projectID)
	}

	// Apply delta updates from raw JSON representation
	if _, ok := raw["title"]; ok && req.Title != nil {
		projectFound["title"] = *req.Title
	}
	if _, ok := raw["done_clean_period"]; ok {
		projectFound["done_clean_period"] = req.DoneCleanPeriod
	}
	if _, ok := raw["git_remote"]; ok {
		if req.GitRemote != nil {
			projectFound["git_remote"] = *req.GitRemote
		} else {
			projectFound["git_remote"] = nil
		}
	}

	// Save to projects registry
	if err := s.fileRepo.WriteProjects(tasksDir, projects); err != nil {
		return nil, fmt.Errorf("failed to write projects registry to disk: %w", err)
	}

	// Resolve database field types from interface maps safely
	var doneCleanPeriod *int
	if projectFound["done_clean_period"] != nil {
		switch v := projectFound["done_clean_period"].(type) {
		case float64:
			val := int(v)
			doneCleanPeriod = &val
		case int:
			val := v
			doneCleanPeriod = &val
		case int64:
			val := int(v)
			doneCleanPeriod = &val
		}
	}

	var gitRemote *string
	if projectFound["git_remote"] != nil {
		if r, ok := projectFound["git_remote"].(string); ok {
			gitRemote = &r
		}
	}

	title, _ := projectFound["title"].(string)

	// Save to DB
	if err := s.dbRepo.Update(ctx, projectID, title, doneCleanPeriod, gitRemote); err != nil {
		return nil, fmt.Errorf("failed to update project in database: %w", err)
	}

	created, _ := projectFound["created_at"].(string)

	return &Response{
		ID:              projectID,
		Title:           title,
		CreatedAt:       created,
		DoneCleanPeriod: doneCleanPeriod,
		GitRemote:       gitRemote,
	}, nil
}

func (s *projectService) DeleteProject(ctx context.Context, tasksDir string, projectID string) error {
	projects, err := s.fileRepo.LoadProjects(tasksDir)
	if err != nil {
		return fmt.Errorf("failed to load projects: %w", err)
	}

	if len(projects) <= 1 {
		return ErrLastProject
	}

	var filtered []map[string]interface{}
	found := false
	for _, p := range projects {
		if p["id"] == projectID {
			found = true
		} else {
			filtered = append(filtered, p)
		}
	}

	if !found {
		return fmt.Errorf("%w: project with ID '%s'", ErrProjectNotFound, projectID)
	}

	// Update Disk File
	if err := s.fileRepo.WriteProjects(tasksDir, filtered); err != nil {
		return fmt.Errorf("failed to update projects registry: %w", err)
	}

	// Delete local task files
	if err := s.fileRepo.DeleteProjectDir(tasksDir, projectID); err != nil {
		return fmt.Errorf("failed to delete project files: %w", err)
	}

	// Cascade delete from DB
	if err := s.dbRepo.Delete(ctx, projectID); err != nil {
		return fmt.Errorf("failed to delete project from database: %w", err)
	}

	return nil
}

File: src/jotter/features/projects/service.py
"""Application service orchestrating Project use cases."""

import sqlite3
from pathlib import Path
from typing import Any, Self

from jotter.features.buckets.repo import BucketRepository
from jotter.features.projects.domain import Project
from jotter.features.projects.repo import ProjectRepository
from jotter.features.projects.schemas import ProjectCreate, ProjectResponse, ProjectUpdate
from jotter.shared.exceptions import EntityNotFoundError, ValidationError
from jotter.shared.slug import slugify


class ProjectApplicationService:
    def __init__(self, project_repo: ProjectRepository, bucket_repo: BucketRepository):
        self.project_repo = project_repo
        self.bucket_repo = bucket_repo

    @classmethod
    def from_data_dir(cls, data_dir: Path | str, conn: sqlite3.Connection) -> Self:
        return cls(
            project_repo=ProjectRepository(data_dir, conn),
            bucket_repo=BucketRepository(data_dir, conn),
        )

    def get_all_projects(self) -> list[ProjectResponse]:
        projects = self.project_repo.get_all()
        if not projects:
            # Seed default project if empty
            default_p = Project.create(name="Default", project_id="default")
            self.project_repo.save(default_p)
            self.bucket_repo.get_all("default")
            projects = [default_p]

        return [self._to_response(p) for p in projects]

    def get_project(self, project_id: str) -> ProjectResponse:
        return self._to_response(self.project_repo.get(project_id))

    def create_project(
        self, req: ProjectCreate, default_buckets: list[dict[str, Any]] | None = None
    ) -> ProjectResponse:
        title = (getattr(req, "title", None) or getattr(req, "name", "")).strip()
        if not title:
            raise ValidationError("Project title cannot be empty")

        slug = getattr(req, "id", None)
        if not slug:
            slug = slugify(title)
        if not slug:
            slug = "project"

        if self.project_repo.exists(slug):
            raise ValidationError(f"Project with ID '{slug}' already exists")

        project = Project.create(
            name=title,
            project_id=slug,
            git_remote=req.git_remote,
            done_clean_period=req.done_clean_period,
        )

        self.project_repo.save(project)
        # Seed default buckets for new project
        self.bucket_repo.get_all(project.id)

        return self._to_response(project)

    def update_project(self, project_id: str, req: ProjectUpdate) -> ProjectResponse:
        project = self.project_repo.get(project_id)

        title = req.title.strip() if req.title is not None else None
        project.update_details(
            name=title,
            git_remote=req.git_remote,
            done_clean_period=req.done_clean_period,
        )

        self.project_repo.save(project)
        return self._to_response(project)

    def delete_project(self, project_id: str) -> None:
        if not self.project_repo.exists(project_id):
            raise EntityNotFoundError(f"Project '{project_id}' not found")
        self.project_repo.delete(project_id)

    def _to_response(self, p: Project) -> ProjectResponse:
        return ProjectResponse(
            id=p.id,
            title=p.name,
            git_remote=p.git_remote,
            done_clean_period=p.done_clean_period,
            created_at=p.created_at,
        )
